pkg/registry/core/service/allocator/bitmap.go

```diff
@@ -205,3 +254,37 @@ func (rss randomScanStrategy) AllocateBit(allocated *big.Int, max, count int) (i
 }
 
 var _ bitAllocator = randomScanStrategy{}
+
+// randomScanReservedStrategy choose a random address from the provided big.Int and then scans
+// forward looking for the next available address. The big.Int range is subdivided so it will try
+// to allocate first from the reserved upper range of addresses (it will wrap the upper subrange if necessary).
+// If there is no free address it will try to allocate one from the lower range too.
+type randomScanReservedStrategy struct {
+	rand     *rand.Rand
+	reserved int
+}
+
+func (rss randomScanReservedStrategy) AllocateBit(allocated *big.Int, max, count int) (int, bool) {
+	if count >= max {
+		return 0, false
+	}
+	// size of the upper subrange, prioritized for random allocation
+	subrangeMax := max - rss.reserved
+	// try to get a value from the upper range [rss.reserved, max]
+	offset := rss.rand.Intn(subrangeMax)
+	for i := 0; i < subrangeMax; i++ {
+		at := rss.reserved + ((offset + i) % subrangeMax)
```

Comment:
Why are we doing this modulo?  Will it ever be > max?  How?

Reply:
worst case
i := subRangeMax
offset := subRangeMax
subrangeMax := max - rss.reserved -> rss.reserved = max - subrangeMax

offset + i = subRangeMax + subRangeMax = 2*max - 2*rss.reserved 
at = rss.reserved + 2*max - 2*rss.reserved = 2*max - rss.reserved , with rss.reserved < max